### ISSUE TITLE
Whisper Tiny: sherpa-onnx produces much higher CER than faster-whisper on the same audio

### ISSUE THREAD

**Issue by zhitao-zeng** [k2-fsa/sherpa-onnx#2900]
---

**Environment**

* OS: Linux
* sherpa-onnx: current master (custom build)
* onnxruntime: onnxruntime.quant
* faster-whisper: latest stable
* Audio sample rate: 16 kHz
* Task: `transcribe`
* Language: `zh`

---

**Reproduction Summary**

Using **the same Whisper Tiny–scale model**, **the same audio**, and **the same reference text**, I observe a **large accuracy gap** between:

* `faster-whisper`
* `sherpa-onnx` (offline Whisper)

The difference is consistent on a Chinese FLEURS sample.

---

**Test Audio**

* Dataset: **FLEURS (zh)**
* Duration: **5.94 s**
* Sample rate: **16 kHz**
* Reference text:

```
现场活动包括打猎、钓鱼、摄影、观鸟、参观公园和研究生态系统。
```

---

## 1. faster-whisper Result (Baseline)

**Model**

```
faster-whisper-tiny-zh-v2
```

**Configuration**

* Device: CUDA
* Compute type: float16
* Beam size: 1
* Temperature: 0.0
* Unicode normalization: NFKC

**Output**

```
现场活动包括打猎钓鱼摄影、关钓参观公园和研究生态系统。
```

**Metrics**

* CER: **0.25**
* RTF: **0.078**

This output is imperfect but mostly semantically correct.

---

## 2. sherpa-onnx Whisper Result

**Model files**

```
encoder.onnx
decoder.onnx
tokens.txt
```

**Recognizer configuration**

```
modelCategory: whisper
task: transcribe
language: zh
decodingMethod: greedy_search
maxActivePaths: 4
featureDim: 80
modelingUnit: bpe
provider: cpu
numThreads: 1
```

**Model metadata (InitEncoder log excerpt)**

```
model_type=whisper-custom
is_multilingual=1
n_mels=80
n_audio_ctx=1500
n_text_ctx=448
n_audio_layer=4
n_text_layer=4
n_vocab=51865
```

**Output**

```
想想活动包括大力的关,相关公和究生
```

**Metrics**

* CER: **0.8095**
* Latency: **1.321 s**

---

## Observations

* Same audio waveform and sample rate
* Same task (`transcribe`)
* Same Whisper Tiny–scale architecture
* Both are multilingual Whisper models
* faster-whisper output is partially correct
* sherpa-onnx output shows **severe token corruption**
* Error is not limited to punctuation or spacing

The CER gap is **~0.25 vs ~0.81**, which is large enough to make sherpa-onnx unusable in this setup.

---

## Questions

1. Is this level of degradation expected for **Whisper Tiny + Chinese** in sherpa-onnx?
2. Are there **recommended decoding settings** (beam search, special tokens, suppression rules) for Whisper models that sherpa-onnx requires to match faster-whisper behavior?
3. Could this be related to:

   * ONNX export differences?
   * Tokenizer / BPE handling?
   * Missing Whisper-specific decoding heuristics (e.g. token suppression, language conditioning)?
4. Is there a **reference Whisper Tiny configuration** in sherpa-onnx that achieves comparable CER to faster-whisper?

---

**Expected Behavior**

Given the same Whisper Tiny model and audio, sherpa-onnx should achieve **roughly comparable CER** to faster-whisper, or at least avoid a >3× CER regression.

I can provide:

* The exact audio file
* A minimal reproduction script
* Full recognizer logs

if that helps with debugging.
Additional Notes

I have already integrated VAD in the pipeline, and the same VAD setup is used consistently across all evaluations.

The language parameter is explicitly fixed to zh during inference (no auto language detection).

The ONNX models are exported using the official scripts provided in this repository, without custom modifications.

The behavior described above is not an isolated case:
I observe the same accuracy gap trend consistently across the entire evaluation dataset.

The example shown in this issue is provided as a minimal representative case, but the conclusion is based on dataset-level evaluation, not a single utterance.

**Comment by zhitao-zeng** [comment]
### Update: Preliminary Root Cause Analysis (Mel Spectrogram & Padding Order)

After further investigation, my current hypothesis is that the large CER regression is caused by a **mismatch in the mel-spectrogram computation pipeline between sherpa-onnx and the original Whisper / faster-whisper implementation**, specifically **the order of padding vs. mel extraction**.

#### Observed difference in preprocessing

**Original Whisper / faster-whisper pipeline:**

1. Load raw waveform
2. **Pad or trim waveform to 30 seconds (480,000 samples at 16 kHz)**
3. Compute log-mel spectrogram
   → Always produces **exactly 3000 mel frames**

This is the behavior implemented in `whisper.audio.log_mel_spectrogram()` and relied upon during both pretraining and fine-tuning.

**Current sherpa-onnx Whisper pipeline (observed behavior):**

1. Load raw waveform
2. Compute log-mel spectrogram on the *original-length audio*
   → Produces a **variable number of frames** (e.g., ~594 frames for a 5.9 s utterance)
3. **Pad the mel spectrogram to 3000 frames**

#### Why this likely matters

Although both approaches result in a 3000-frame input tensor, they are **not equivalent**:

* Padding **before** mel computation preserves the temporal statistics expected by Whisper
* Padding **after** mel computation changes:

  * Attention alignment
  * Effective receptive field near the end of the utterance
  * The distribution of silent frames seen by the encoder
* Whisper was trained assuming **waveform-level padding**, not mel-level padding

For short utterances (e.g., FLEURS samples), this discrepancy is systematic and may explain:

* Severe token corruption
* Cross-language hallucination
* > 3× CER regression compared to faster-whisper

#### Evidence

* Same ONNX model
* Same audio waveform
* Same decoding method (greedy)
* Same language fixed to `zh`
* The CER gap appears consistently across the dataset, not just in a single example

This suggests a **systematic preprocessing mismatch** rather than random decoding noise.

#### Next steps (proposed)

* Confirm whether sherpa-onnx can support waveform-level padding **before** mel extraction for Whisper models
* Or document that current Whisper support deviates from the original Whisper preprocessing assumptions
* I can provide a minimal repro script that compares:

  * (A) waveform-pad → mel
  * (B) mel → pad
    using the same encoder.onnx

I believe aligning the mel extraction pipeline with the original Whisper implementation is a necessary condition to achieve comparable CER to faster-whisper.
